specs/Gem.spec

```diff
@@ -0,0 +1,88 @@
+methods {
+    balanceOf(address)         returns(uint) envfree
+    allowance(address,address) returns(uint) envfree
+    totalSupply()              returns(uint) envfree
+}
+
+rule basicTransferSpec {
+     address recip; uint amount;
+
+    env e;
+    address sender = e.msg.sender;
+    
+    require(balanceOf(e.msg.sender) + balanceOf(recip) < totalSupply());
+
+    // mathint type that represents an integer of any size;
+    mathint balance_sender_before = balanceOf(sender);
+    mathint balance_recip_before = balanceOf(recip);
+
+    transfer(e, recip, amount);
+
+    mathint balance_sender_after = balanceOf(sender);
+    mathint balance_recip_after = balanceOf(recip);
+
+    // operations on mathints can never overflow or underflow. 
+    assert recip != sender => balance_sender_after == balance_sender_before - amount,
+        "transfer must decrease sender's gem balance by amount sent";
+
+    assert recip != sender => balance_recip_after == balance_recip_before + amount,
+        "transfer must increase recipient's gem balance by amount sent";
+
+    assert recip == sender => balance_sender_after == balance_sender_before,
+        "transfer must not change sender's gem balance when recip is self";
+}
+
+rule basicMintSpec {
+        address recip; uint amount;
+    
+    env e;
+    require(balanceOf(e.msg.sender) + balanceOf(recip) < totalSupply());
+    address ward = e.msg.sender;
+    mathint balance_recip_before = balanceOf(recip);
+    mathint total_supply_before = totalSupply();
+
+    ward(e, ward, true);
+    mint(e, recip, amount);
+
+    mathint balance_recip_after = balanceOf(recip);
+    mathint total_supply_after = totalSupply();
+
+    assert balance_recip_after == balance_recip_before + amount, "recip balance did not increase on mint";
+    assert total_supply_after == total_supply_before + amount, "total supply did not increase correctly";
+
+}
+
+rule mintMustNeverOverflow {
+        address recip; uint amount;
+    
+    env e;
+    require(totalSupply() + amount > max_uint256);
```

Comment:
maybe better to use mathint here?  totalSupply() + amount is a uint and should revert on overflow, so prover shouldn't consider that path

Reply:
or not revert because it's not a contract but prover just doesn't consider overflowing ops

```

In CVL, arithmetic operators (+, -, * and /) are overloaded: they could mean a machine-arithmetic operation that can overflow, or a mathematical operation that does not overflow. The default interpretation used in almost all cases is the mathematical one. Therefore, the assertion below holds:

uint x;
assert x + 1 > x;

```